pam/pam_adsys.c

```diff
@@ -31,295 +31,261 @@
-{
-	int retval;
-	retval = pam_info(pamh, "Applying user settings");
-	if (retval != PAM_SUCCESS) {
-		return retval;
-	}
-
-	if (memcmp(krb5ccname, (const char *)"FILE:", 5) == 0) {
-		krb5ccname += 5;
-	}
-
-	char **arggv;
-	arggv = calloc(6, sizeof(char *));
-	if (arggv == NULL) {
-		return PAM_BUF_ERR;
-	}
-
-	arggv[0] = "/sbin/adsysctl";
-	arggv[1] = "update";
-	arggv[2] = (char *)(username);
-	arggv[3] = (char *)(krb5ccname);
-	arggv[4] = NULL;
-	if (debug) {
-		arggv[4] = "-vv";
-		arggv[5] = NULL;
-	}
-
-	pid_t pid = fork();
-	if (pid == -1) {
-		pam_syslog(pamh, LOG_ERR, "Failed to fork process");
-		return PAM_SYSTEM_ERR;
-	}
-
-	if (pid > 0) {		/* parent */
-		pid_t retval;
-		int status = 0;
-
-		while ((retval = waitpid(pid, &status, 0)) == -1
-		       && errno == EINTR) ;
-
-		if (retval == (pid_t) - 1) {
-			pam_syslog(pamh, LOG_ERR,
-				   "waitpid returns with -1: %m");
-			free(arggv);
-			return PAM_SYSTEM_ERR;
-		} else if (status != 0) {
-			if (WIFEXITED(status)) {
-				pam_syslog(pamh, LOG_ERR,
-					   "adsysctl update %s %s failed: exit code %d",
-					   username, krb5ccname,
-					   WEXITSTATUS(status));
-			} else if (WIFSIGNALED(status)) {
-				pam_syslog(pamh, LOG_ERR,
-					   "adsysctl update %s %s failed: caught signal %d%s",
-					   username, krb5ccname,
-					   WTERMSIG(status),
-					   WCOREDUMP(status) ? " (core dumped)"
-					   : "");
-			} else {
-				pam_syslog(pamh, LOG_ERR,
-					   "adsysctl update %s %s failed: unknown status 0x%x",
-					   username, krb5ccname, status);
-			}
-			free(arggv);
-			return PAM_CRED_ERR;
-		}
-		free(arggv);
-		return PAM_SUCCESS;
-
-	} else {		/* child */
-		if (debug) {
-			pam_syslog(pamh, LOG_DEBUG, "Calling %s ...", arggv[0]);
-		}
-
-		execv(arggv[0], arggv);
-		int i = errno;
-		pam_syslog(pamh, LOG_ERR, "execv(%s,...) failed: %m", arggv[0]);
-		free(arggv);
-		_exit(i);
-	}
-
-	return PAM_SYSTEM_ERR;	/* will never be reached. */
+static int update_policy(pam_handle_t *pamh, const char *username, const char *krb5ccname, int debug) {
+    int retval;
+    retval = pam_info(pamh, "Applying user settings");
+    if (retval != PAM_SUCCESS) {
+        return retval;
+    }
+
+    if (memcmp(krb5ccname, (const char *)"FILE:", 5) == 0) {
+        krb5ccname += 5;
+    }
+
+    char **arggv;
+    arggv = calloc(6, sizeof(char *));
+    if (arggv == NULL) {
+        return PAM_BUF_ERR;
+    }
+
+    arggv[0] = "/sbin/adsysctl";
+    arggv[1] = "update";
+    arggv[2] = (char *)(username);
+    arggv[3] = (char *)(krb5ccname);
+    arggv[4] = NULL;
+    if (debug) {
+        arggv[4] = "-vv";
+        arggv[5] = NULL;
+    }
+
+    pid_t pid = fork();
+    if (pid == -1) {
+        pam_syslog(pamh, LOG_ERR, "Failed to fork process");
+        return PAM_SYSTEM_ERR;
+    }
+
+    if (pid > 0) { /* parent */
+        pid_t retval;
+        int status = 0;
+
+        while ((retval = waitpid(pid, &status, 0)) == -1 && errno == EINTR)
+            ;
```

Comment:
That was my second point, I’m still worried about this one :)

Reply:
I'm not sure... This kind of formatting looks kind of familiar for these types of blocks without inside code. We can add a `continue` there to make sure the formatter behaves, wdyt? Might even make the code clearer

Reply:
as you wish, look for what is the easiest. It could just be buming to 160 characters :)

Reply:
Oh, no. The limit was fine. I think it uses the newline for the `;` just to make it clear that it is a `do nothing` line

Reply:
soooooooooo ugly, can we try `while ((retval = waitpid(pid, &status, 0)) == -1 && errno == EINTR) {};` ?

Reply:
yes, because when I see "continue" I think of getting into the loop in a busy one, while actually, the while is blocking and never gets into `{}`

Reply:
This is what clang-format has to say about it:
```C
while ((retval = waitpid(pid, &status, 0)) == -1 && errno == EINTR) {
};
```

Reply:
hum, not that bad, wdyt?

Reply:
It still hurts, but it's definitely better than a random indented `;` in the middle of the code

Reply:
agreed

Reply:
Honestly, I'm just glad I get to not look at
```C
int function()
{
}
```
anymore